platform/ui/src/components/downloadDialog/DownloadDialog.js

```diff
@@ -0,0 +1,235 @@
+import './DownloadDialog.styl';
+
+import React, { PureComponent } from 'react';
+import { withTranslation } from '../../utils/LanguageProvider';
+import Modal from 'react-bootstrap-modal';
+import 'react-bootstrap-modal/lib/css/rbm-patch.css';
+import PropTypes from 'prop-types';
+
+class DownloadDialog extends PureComponent {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      isOpen: this.props.isOpen,
+      width: 3000,
+      height: 3000,
+      fileName: '',
+      fileType: 'png',
+      showAnnotations: true,
+      fileTypeOptions: [
+        {
+          key: 'jpg',
+          value: 'jpg'
+        },
+        {
+          key: 'png',
+          value: 'png'
+        }
+      ],
+      previewElementRef: null
+    };
+
+    this.submitForm = this.submitForm.bind(this);
+    this.onClose = this.onClose.bind(this);
+  }
+
+  submitForm(e) {
+    e.preventDefault();
+
+    const formData = {
+      width: this.state.width,
+      height: this.state.height,
+      fileName: this.state.fileName,
+      fileType: this.state.fileType,
+      showAnnotations: this.state.showAnnotations,
+    };
+
+    const thereIsAnyEmpty = Object.values(formData).some(field => field === '');
+
+    if (thereIsAnyEmpty) {
+      alert('Please full fill all fields.');
+      return false;
+    }
+
+    this.props.takeAndDownloadSnapShot(formData);
+
+    return false;
+  };
+
+  static propTypes = {
+    isOpen: PropTypes.bool.isRequired,
+    toggleDownloadDialog: PropTypes.func.isRequired,
+    takeAndDownloadSnapShot: PropTypes.func.isRequired,
+    mountPreview: PropTypes.func.isRequired,
+    onResize: PropTypes.func.isRequired,
+    toggleAnnotations: PropTypes.func.isRequired,
+    forceRenderUpdate: PropTypes.func.isRequired,
+    setCacheReferences: PropTypes.func.isRequired,
+    cleanViewPortClone: PropTypes.func.isRequired,
+  };
+
+  /**
+   * assignRef
+   * This way to pass trough a ref is needed to make a dom ref accessible on
+   * static method getDerivedStateFromProps, what will replace componentWillReceiveProps
+   * See https://fb.me/react-async-component-lifecycle-hooks for details.
+   * @param references dom element ref to cache on state
+   */
+  assignRef = references => {
+    // Control recalls, allowing this instance on state only once
+    if (this.props.updateHash === null && references !== null) {
+      this.setState({ previewElementRef: references,  });
+
+      this.props.setCacheReferences(
```

Comment:
What is `setCacheReferences`?

Reply:
Every instance of the Download Engine caches the dom stuff to prevent new query selections and gets from Cornerstone. Improves performance and code readability, once inside the Engine Class we can use them as `this.$activeElement` instead of

`
const originalElement = cornerstone.getEnabledElement(this.$activeViewport).element;
const originalCanvas = originalElement.querySelector('canvas');
`